### DISCUSSION TITLE
[Request] Ravynos

### DISCUSSION THREAD

**Discussion by jhay06** [ravynsoft/ravynos#309] [Q&A] [answered]
Hi ,

I love the project and idea, it looks like it's the same idea as darling on linux

i would like to help with this project , but before that can i ask ? what tools or IDE did you use and how can i debug this without installing  ? 

Thanks

**Comment by mszoek** [comment, marked as answer]
Hi!

Yes, very similar to Darling in several ways. I like to think of ravynOS as akin to ReactOS but for Mac.

Help is totally welcome! At this stage, ravynOS is self-hosting... you need to use it to build it. Xcode on Mac can be used to build applications that build on ravynOS with a little porting work (Makefiles vs pbxproj, compiling & copying binary NIB files). They can also be written by hand without NIBs if you know how Cocoa works. Most of what I've done has been in tmux + vim on the running system (where I am right now!).

If you are working on low-level or CLI stuff, a VM will work. It has everything except a GUI because we don't yet have video drivers for virtual GPUs. To work on apps or UI, you probably do need to install it on some real hardware: Intel HD/Iris series, AMD embedded, and Radeon GPUs are currently supported.

**Reply by jhay06** [reply]
I see ,  so some apps that was built with mac can/might work on ravyn os ? 
if i try to help with the project and i focused on running the mac apps ? what files from the your source code should i look for and debug. 

I don't know where to start ,  but i know i can help with the project if i know what files should i focused to debug mac apps on ravyn os and modified that file to make the app works on ravyn thanks 

sorry for my bad english

**Reply by mszoek** [reply]
The goal is to run unmodified Mac apps. We're not there yet - there is a lot of plumbing to create first, and that's where most of my effort is going. The main milestones to get there are: essential frameworks implementing all the required functions, Mach-O dynamic linker and loader, converting the rest of the OS to Mach-O from ELF, and building the necessary services and environments like Finder, SystemUIServer, etc so that apps can actually function.

**Comment by jhay06** [comment]
hi @mszoek . debuging and fixing some partial code ? does it need to rebuild the whole OS?   

hope there's a guide for how to if it doesn't need to build the whole OS for just a simple 1 line code for example 

I really want to help , but i still dont know how to help since building the Whole OS is a very time consuming , but if it doesnt require to build the whole OS , can you  add some ReadME for the new contributor ? 

example if i changed the cocoa part , then build it where should i put it from running system ? or how to debug it first before putting it to the system , as well as other dependency of the OS . 

than you very much

**Reply by mszoek** [reply]
The OS builds in 3 stages called "kernel", "base", and "system". Kernel is the kernel, obviously. Base is the full BSD base system with all libraries, tools, etc. This is the longest stage. System is the top layer with Frameworks, CoreServices, WindowServer, Dock, etc.

If you are changing Cocoa, you can do `./tools/ravynOS/build.sh system` from the top level of the source tree to build just the top layer. It's fairly fast once you have built it the first time. You can also usually do `make` inside of individual directories to build just that part of the source tree (e.g. `lib/libc`). I don't recommend this for the `system` targets but it generally works for `base` targets.

Once you've built, you can do `./tools/ravynOS/build.sh iso` to make a new ISO or you can use `./tools/ravynOS/build.sh install` to update the running system with any changes in the object tree (be careful). You can also use `make install` in specific directories to update parts of the system. Pass `DESTDIR=/xxx` to the `make install` if you want to use `/xxx` as the root directory... this lets you install everything to a jail for testing.

**Comment by ghost** [comment]
Hi, could you help me could you direct me to the right page from where to download the iso of ravynOS x86 ... thanks

**Reply by dleggo** [reply]
You can get the ISO(s) from the [releases page](https://github.com/ravynsoft/ravynos/releases). Currently there are only ```amd64``` releases.
